brave/brave-core
# Support Unstoppable Domains via DNS over HTTPS behind a feature flag

This PR adds support of Unstoppable Domains via public DoH endpoint https://resolver.unstoppable.io/dns-query provided by Unstoppable Domains, which currently is just a redirect to Cloudflare server https://eth.resolver.cloudflare-eth.com/dns-query without any logging.

This endpoint will only be used for name resolution if user set the resolve method to `Public DNS over HTTPS server`, and it will only be used for name resolution of domains ended with `.crypto`. The way we achieve this in the PR is hooking in where chromium updates the secure dns configuration in network service, add the custom servers if pref is set to public DoH. These added custom servers will be skipped for non-crypto domains before making actual DNS attempt in DNS transactions.

An interstitial page will be shown to users to opt in when users try to load *.crypto domains.
<img width="1051" alt="Screen Shot 2021-02-10 at 8 56 49 AM" src="https://user-images.githubusercontent.com/4730197/107546382-258ec980-6b81-11eb-9851-bd8c88a9b2cc.png">

New setting for resolve method is added:
<img width="711" alt="Screen Shot 2021-02-10 at 9 25 34 AM" src="https://user-images.githubusercontent.com/4730197/107547375-38ee6480-6b82-11eb-8deb-e846422323f2.png">

Resolves https://github.com/brave/brave-browser/issues/14099

## Submitter Checklist:


## Reviewer Checklist:


## After-merge Checklist:

  changes has landed on.

## Test Plan:
1. Open Brave and visit brave.crypto, it should show error page saying `This site can’t be reached`.
<img width="1054" alt="Screen Shot 2021-02-10 at 8 53 59 AM" src="https://user-images.githubusercontent.com/4730197/107542943-916f3300-6b7d-11eb-9406-1fdc47213cd2.png">

2. Go to settings, search for unstoppable, there should be no unstoppable domains resolve method setting.

3. Go to brave://flags and search for unstoppable and enable the feature flag.
<img width="1066" alt="Screen Shot 2021-02-10 at 8 54 09 AM" src="https://user-images.githubusercontent.com/4730197/107542973-9af89b00-6b7d-11eb-9dda-88f5ddec72d4.png">

4. Restart browser and go to brave.crypto should now should the interstitial for users to opt-in unstoppable domains support.
<img width="1051" alt="Screen Shot 2021-02-10 at 8 56 49 AM" src="https://user-images.githubusercontent.com/4730197/107543286-ee6ae900-6b7d-11eb-9308-21994dbcc2a5.png">

5. Click Disable, should now see site can't be reached and `Method to resolve unstoppable domains` in brave://settings should now be `Disabled`.

6. Change the resolve method setting back to ask and visit brave.crypto again.
<img width="696" alt="Screen Shot 2021-01-26 at 9 27 03 PM" src="https://user-images.githubusercontent.com/4730197/106043111-ce71fa80-6092-11eb-884e-8f040c707038.png">

7. Opt-in interstitial should be shown again, and this time click `Proceed`.

8. Page should be loaded now. (Note that there will be a brief moment that it is showing site is not reached while the page is loading, this is a known and will be addressed in a new issue.)

9. `Method to resolve unstoppable domains` in brave://settings should now be `Public DNS over HTTP server`.
10. Open a new tab and visit brad.crypto, page should be loaded normally without any interstitial pages.
